[iOS] Put the UIScrollView into a UIView in a border handler

kubaflo opened the pull request:
### Description of Change

Putting UIScroll view in a UIView is necessary when applying masking which is a case with the border control. However, checking from border handler is better than checking up from UiScrollViews like in here https://github.com/dotnet/maui/pull/14096

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes https://github.com/dotnet/maui/issues/8923
Fixes https://github.com/dotnet/maui/issues/14091
Fixes https://github.com/dotnet/maui/issues/7524
Fixes https://github.com/dotnet/maui/issues/14091
Fixes https://github.com/dotnet/maui/issues/20736
Fixes https://github.com/dotnet/maui/issues/18486
Closes https://github.com/dotnet/maui/pull/24683

jsuarezruiz commented:
/azp run

PureWeen commented on src/Core/src/Handlers/Border/BorderHandler.iOS.cs:
```diff
@@ -51,8 +52,18 @@ static partial void UpdateContent(IBorderHandler handler)
+				// If the content is a UIScrollView, we need a container to handle masks and clip shapes effectively.
+				if (platformContent is UIScrollView && content.Handler is not null)
+				{
+					content.Handler.HasContainer = true;
```
I think with how HasContainer works this might be risky

Something might cause Container to reset back to false

![image](https://github.com/user-attachments/assets/cc77bad8-64c8-4c61-b9b5-9d608a423cca)

I would be curious at some point to explore a way for the parent to flag the child as needing a container but for now I think we'll just have to add a view here and then add the content to that view.

rmarinho commented:
/azp run

maonaoda commented:
This fix seems to cause the Border size to change strangely in certain situations (backstage back to the foreground, page deleted).
![截屏2024-10-29 12 48 18](https://github.com/user-attachments/assets/58f8aca3-dd82-45dd-8f78-1e388e00c4a2)

jsuarezruiz commented:
/rebase

jsuarezruiz commented:
> This fix seems to cause the Border size to change strangely in certain situations (background back to the foreground, page deleted). The Green one is the containerView and The Red is Border`s Stroke
> 
> ![截屏2024-10-29 13 14 41](https://private-user-images.githubusercontent.com/32831595/380983674-dc8537f5-6cda-4d20-8322-8dbaaf6acb73.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MjI2MzUsIm5iZiI6MTczMTQyMjMzNSwicGF0aCI6Ii8zMjgzMTU5NS8zODA5ODM2NzQtZGM4NTM3ZjUtNmNkYS00ZDIwLTgzMjItOGRiYWFmNmFjYjczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDE0Mzg1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5NDdhODIyZTMwOTIyZDQ5NDgyYjQzNjUxN2NmMDFlMWU2MDIzZjM2ZDdhZDMwZmU0Yzg1MWZjMzMwNDNkMmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eO5mGpEYOKKrJbTYLEAq2vZW8SgcagkPOb4-6iFJYB4)

Could you share a piece of XAML or C# to reproduce your UI and add an UITest to validate it?

kubaflo commented:
> This fix seems to cause the Border size to change strangely in certain situations (background back to the foreground, page deleted). The Green one is the containerView and The Red is Border`s Stroke
> 
> ![截屏2024-10-29 13 14 41](https://private-user-images.githubusercontent.com/32831595/380983674-dc8537f5-6cda-4d20-8322-8dbaaf6acb73.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0Mzc2NjEsIm5iZiI6MTczMTQzNzM2MSwicGF0aCI6Ii8zMjgzMTU5NS8zODA5ODM2NzQtZGM4NTM3ZjUtNmNkYS00ZDIwLTgzMjItOGRiYWFmNmFjYjczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDE4NDkyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4ZmNkOGI3ZThlOWY4MDg5ODY3OTQ3YzEyN2FhZWI3YWVhOTJhZjcyZWZlYmViYzhkNzMzMDIyOTc2N2Y5M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._PqYnYhlXjzF20PCnLhzxxbynlCKkyR1Cn9aKTudJxk)

As @jsuarezruiz wrote could you please share a code with us, so we can try to replicate it? Also, I've modified this PR a bit so it would be cool if you try again and let us know the outcome. Thanks!

PureWeen commented:
/azp run

PureWeen commented:
/azp run

maonaoda commented:
fine, I will confirm.

maonaoda commented:
Code: 
[TestMauiApp.zip](https://github.com/user-attachments/files/17728303/TestMauiApp.zip)

The problem now is very serious...

9.0GA without this PR：

https://github.com/user-attachments/assets/d44b1be2-c3c0-43f7-ae37-d89565ab9aef

9.0GA with this PR：

https://github.com/user-attachments/assets/3149edcb-1749-495c-92a7-d333c1bdbfc3

maonaoda commented:
Adding NeedsContainer to EditorHandler may be can solve this issue.

kubaflo commented:
> Code: [TestMauiApp.zip](https://github.com/user-attachments/files/17728303/TestMauiApp.zip)
> 
> The problem now is very serious...
> 
> 9.0GA without this PR：
> 
>  2024-11-13.15.05.59.mov 
> 
> 9.0GA with this PR：
> 
>  2024-11-13.15.29.22.mov

Could you try again? I removed constraints and replaced them with frame

maonaoda commented:
Great, that's it.

https://github.com/user-attachments/assets/969010de-1547-4dad-b235-0f954f2652d0

※The newest code is on https://github.com/maonaoda/TestMaui/tree/bug/25817
Although having issue #25817,
but the issue this PR concerns has been confirmed to be resolved.
Thanks a lot :)

---
Other than this, I have determined that `Adding NeedsContainer to EditorHandler ` also is possible to fix.

kubaflo commented:
@maonaoda 
Cool! Yes, adding a needs container would fix it but we wanted to find a better solution that would avoid editor handler to have check if it is wrapped inside a border. Additionally, this would require having the same code for all the views that inherit from UiScrollView

Thanks for verifying the current code!

PureWeen commented:
/azp run

jsuarezruiz commented on src/Core/src/Handlers/Border/BorderHandler.iOS.cs:
```diff
@@ -51,8 +48,24 @@ static partial void UpdateContent(IBorderHandler handler)
-				platformView.AddSubview(platformContent);
+
+				// If the content is a UIScrollView, we need a container to handle masks and clip shapes effectively.
+				if (platformContent is UIScrollView)
```
Could we create an UITest and verify the Border after update editors using https://github.com/maonaoda/TestMaui/tree/bug/25817?

rmarinho reviewed (dismissed):
Can we rebase to make sure we have latest changes, and add a test?

kubaflo commented:
@rmarinho I've rebased and added a UITest

PureWeen commented:
/azp run

PureWeen commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue20736.cs:
```diff
@@ -0,0 +1,25 @@
+		[Test]
+		[Category(UITestCategories.Editor)]
+		[Category(UITestCategories.Border)]
+		public void EditorScrollingWhenEnclosedInBorder()
```
Looks like this is failing on WinUI

kubaflo replied:
Should work now

PureWeen commented:
/azp run

PureWeen commented:
/rebase

PureWeen commented:
/azp run

PureWeen commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue20736.cs:
```diff
@@ -0,0 +1,25 @@
+		[Test]
+		[Category(UITestCategories.Editor)]
+		[Category(UITestCategories.Border)]
+		public void EditorScrollingWhenEnclosedInBorder()
```
It looks like this is a little bit flakey on Android

https://dev.azure.com/xamarin/public/_build/results?buildId=131616&view=ms.vss-test-web.build-test-results-tab&runId=3552379&resultId=100044&paneView=debug

kubaflo replied:
Yea, it the screenshots look different on various android devices. Maybe I should disable Android for this?

PureWeen commented:
/azp run

PureWeen commented on src/Core/src/PublicAPI/net-ios/PublicAPI.Shipped.txt:
```diff
@@ -2141,7 +2141,6 @@ override Microsoft.Maui.Handlers.ScrollViewHandler.ConnectHandler(UIKit.UIScroll
 override Microsoft.Maui.Handlers.ScrollViewHandler.CreatePlatformView() -> UIKit.UIScrollView!
 override Microsoft.Maui.Handlers.ScrollViewHandler.DisconnectHandler(UIKit.UIScrollView! platformView) -> void
 override Microsoft.Maui.Handlers.ScrollViewHandler.GetDesiredSize(double widthConstraint, double heightConstraint) -> Microsoft.Maui.Graphics.Size
-override Microsoft.Maui.Handlers.ScrollViewHandler.NeedsContainer.get -> bool
```
Can you use *REMOVE* here inside the Unshipped.text files?

kubaflo committed:
Put the UIScrollView into a generalWrapperView in a border handler

PureWeen commented:
/azp run

PureWeen commented:
- failing tests are unrelated
